File: dotnet-mcp/McpServer/Controllers/McpController.cs
using System.Text;
using System.Text.Json;
using Microsoft.AspNetCore.Mvc;
using McpServer.Models;
using McpServer.Services;

namespace McpServer.Controllers;

[ApiController]
[Route("v1")]
public class McpController : ControllerBase
{
    private readonly SessionService _sessionService;
    private readonly ILogger<McpController> _logger;
    
    public McpController(SessionService sessionService, ILogger<McpController> logger)
    {
        _sessionService = sessionService;
        _logger = logger;
    }

    // MCP initialize endpoint for protocol compatibility
    [HttpPost("initialize")]
    public IActionResult Initialize([FromBody] object request)
    {
        // Return a minimal MCP-compliant response
        var response = new
        {
            id = Guid.NewGuid().ToString(),
            object_type = "mcp-initialize-response",
            server = new {
                name = ".NET MCP Example Server",
                version = "1.0.0",
                description = "A .NET MCP server supporting stat analysis and text completions."
            },
            capabilities = new[] { "text-completions", "stat-analysis", "streaming" }
        };
        return Ok(response);
    }
    
    // Session creation endpoint
    [HttpPost("session")]
    public IActionResult CreateSession([FromBody] SessionCreateRequest request)
    {
        try
        {
            _logger.LogInformation("Session creation requested with ID: {Id}", request.Id);
            
            var sessionId = Guid.NewGuid().ToString();
            
            if (_sessionService.TryCreateSession(sessionId, request.Attributes.AccessToken, out var sessionInfo))
            {
                var response = new SessionCreateResponse
                {
                    Id = request.Id,
                    SessionId = sessionId
                };
                
                _logger.LogInformation("Session created successfully: {SessionId}", sessionId);
                return Ok(response);
            }
            else
            {
                return StatusCode(500, CreateErrorResponse(request.Id, "internal_error", "Failed to create session"));
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating session");
            return StatusCode(500, CreateErrorResponse(request.Id, "internal_error", "An unexpected error occurred"));
        }
    }
    
    // Text completion endpoint - Streamable HTTP
    [HttpPost("text/completions")]
    public IActionResult GetTextCompletions([FromBody] TextCompletionRequest request)
    {
        try
        {
            _logger.LogInformation("Text completion/stat analysis requested for session {SessionId}", request.SessionId);

            if (!_sessionService.TryGetSession(request.SessionId, out var sessionInfo))
            {
                _logger.LogWarning("Invalid session ID: {SessionId}", request.SessionId);
                return BadRequest(CreateErrorResponse(request.Id, "invalid_session", "Invalid or expired session ID"));
            }

            // If stats are present, do deterministic stat analysis
            if (request.Inputs?.Stats != null)
            {
                var result = new Dictionary<string, string>();
                var p1 = request.Inputs.Stats.Player1;
                var p2 = request.Inputs.Stats.Player2;
                foreach (var stat in p1.Keys)
                {
                    if (p2.ContainsKey(stat))
                    {
                        if (p1[stat] > p2[stat]) result[stat] = "player1";
                        else if (p1[stat] < p2[stat]) result[stat] = "player2";
                        else result[stat] = "tie";
                    }
                }
                // Save to session history
                sessionInfo.AnalysisHistory.Add(new Dictionary<string, string>(result));
                var response = new
                {
                    id = request.Id,
                    type = "stat-analysis-response",
                    result,
                    history = sessionInfo.AnalysisHistory
                };
                return Ok(response);
            }
            // If prompt is present, do text completion (demo/dummy response)
            else if (!string.IsNullOrWhiteSpace(request.Inputs?.Prompt))
            {
                // Demo: echo the prompt with a canned completion
                var completion = $"{request.Inputs.Prompt} [COMPLETION]";
                var response = new
                {
                    id = request.Id,
                    type = "text-completion-response",
                    result = new { text = completion },
                    usage = new { prompt_tokens = request.Inputs.Prompt.Length, completion_tokens = 1, total_tokens = request.Inputs.Prompt.Length + 1 }
                };
                return Ok(response);
            }
            else
            {
                return BadRequest(CreateErrorResponse(request.Id, "invalid_request", "Missing stats or prompt in request."));
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing text completion/stat analysis");
            return StatusCode(500, CreateErrorResponse(request.Id, "internal_error", "An unexpected error occurred"));
        }
    }
    
    // Legacy HTTP+SSE streaming endpoint
    [HttpPost("text/completions/stream")]
    public IActionResult StreamTextCompletions([FromBody] TextCompletionRequest request)
    {
        try
        {
            _logger.LogInformation("SSE text completion/stat analysis requested for session {SessionId}", request.SessionId);

            if (!_sessionService.TryGetSession(request.SessionId, out var sessionInfo))
            {
                _logger.LogWarning("Invalid session ID: {SessionId}", request.SessionId);
                return BadRequest(CreateErrorResponse(request.Id, "invalid_session", "Invalid or expired session ID"));
            }

            // Set up SSE headers
            Response.Headers["Content-Type"] = "text/event-stream";
            Response.Headers["Cache-Control"] = "no-cache";
            Response.Headers["Connection"] = "keep-alive";

            // If stats are present, stream stat analysis
            if (request.Inputs?.Stats != null)
            {
                var result = new Dictionary<string, string>();
                var p1 = request.Inputs.Stats.Player1;
                var p2 = request.Inputs.Stats.Player2;
                foreach (var stat in p1.Keys)
                {
                    if (p2.ContainsKey(stat))
                    {
                        if (p1[stat] > p2[stat]) result[stat] = "player1";
                        else if (p1[stat] < p2[stat]) result[stat] = "player2";
                        else result[stat] = "tie";
                    }
                }
                // Save to session history
                sessionInfo.AnalysisHistory.Add(new Dictionary<string, string>(result));
                return new StreamingResponse(async stream =>
                {
                    var writer = new StreamWriter(stream);
                    // Stream each stat as a separate SSE event
                    foreach (var stat in result.Keys)
                    {
                        var chunk = new
                        {
                            id = request.Id,
                            type = "stat-analysis-stream-response",
                            stat,
                            winner = result[stat]
                        };
                        var json = JsonSerializer.Serialize(chunk);
                        await writer.WriteAsync($"data: {json}\n\n");
                        await writer.FlushAsync();
                        await Task.Delay(100); // Simulate streaming
                    }
                    // Final event with full history
                    var final = new
                    {
                        id = request.Id,
                        type = "stat-analysis-stream-history",
                        history = sessionInfo.AnalysisHistory
                    };
                    var finalJson = JsonSerializer.Serialize(final);
                    await writer.WriteAsync($"data: {finalJson}\n\n");
                    await writer.WriteAsync("data: [DONE]\n\n");
                    await writer.FlushAsync();
                });
            }
            // If prompt is present, stream a text completion (demo)
            else if (!string.IsNullOrWhiteSpace(request.Inputs?.Prompt))
            {
                var prompt = request.Inputs.Prompt;
                return new StreamingResponse(async stream =>
                {
                    var writer = new StreamWriter(stream);
                    // Simulate streaming the completion one word at a time
                    var words = (prompt + " [COMPLETION]").Split(' ');
                    foreach (var word in words)
                    {
                        var chunk = new
                        {
                            id = request.Id,
                            type = "text-completion-stream-response",
                            text = word
                        };
                        var json = JsonSerializer.Serialize(chunk);
                        await writer.WriteAsync($"data: {json}\n\n");
                        await writer.FlushAsync();
                        await Task.Delay(100); // Simulate streaming
                    }
                    await writer.WriteAsync("data: [DONE]\n\n");
                    await writer.FlushAsync();
                });
            }
            else
            {
                return BadRequest(CreateErrorResponse(request.Id, "invalid_request", "Missing stats or prompt in request."));
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing SSE text completion/stat analysis");
            return StatusCode(500, CreateErrorResponse(request.Id, "internal_error", "An unexpected error occurred"));
        }
    }
    
    // Close session endpoint
    [HttpDelete("session/{sessionId}")]
    public IActionResult CloseSession(string sessionId)
    {
        try
        {
            _logger.LogInformation("Session close requested: {SessionId}", sessionId);
            
            if (_sessionService.TryCloseSession(sessionId))
            {
                var response = new SessionCloseResponse
                {
                    Id = Guid.NewGuid().ToString()
                };
                
                _logger.LogInformation("Session closed successfully: {SessionId}", sessionId);
                return Ok(response);
            }
            else
            {
                _logger.LogWarning("Failed to close nonexistent session: {SessionId}", sessionId);
                return NotFound(CreateErrorResponse(Guid.NewGuid().ToString(), "invalid_session", "Session not found"));
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error closing session");
            return StatusCode(500, CreateErrorResponse(Guid.NewGuid().ToString(), "internal_error", "An unexpected error occurred"));
        }
    }
    
    // Helper methods
    private static ErrorResponse CreateErrorResponse(string requestId, string code, string message)
    {
        return new ErrorResponse
        {
            Id = requestId,
            Error = new McpError
            {
                Code = code,
                Message = message
            }
        };
    }
    
    // The old GenerateDemoResponses method is now obsolete and not used. It is safe to remove it.
}

// Custom result type for SSE streaming
public class StreamingResponse : IActionResult
{
    private readonly Func<Stream, Task> _callback;
    
    public StreamingResponse(Func<Stream, Task> callback)
    {
        _callback = callback;
    }
    
    public async Task ExecuteResultAsync(ActionContext context)
    {
        await _callback(context.HttpContext.Response.Body);
    }
}

File: docker-mcp/src/controllers/McpController.js
/**
 * MCP Controller - Handles MCP protocol endpoints
 */
const { v4: uuidv4 } = require('uuid');
const sessionService = require('../services/SessionService');
const logger = require('../services/LoggerService');
const { 
  SessionCreateResponse, 
  TextCompletionStreamResponse, 
  TextCompletionDelta,
  TextCompletionUsage,
  SessionCloseResponse,
  ErrorResponse 
} = require('../models/McpModels');

/**
 * Create a new session
 */
function createSession(req, res) {
  try {
    const request = req.body;
    logger.info(`Session creation requested with ID: ${request.id}`);
    
    // Generate a unique session ID
    const sessionId = uuidv4();
    
    // Create the session
    const accessToken = request.attributes?.access_token;
    sessionService.createSession(sessionId, accessToken);
    
    // Create and send the response
    const response = new SessionCreateResponse(request.id, sessionId);
    
    logger.info(`Session created successfully: ${sessionId}`);
    return res.json(response);
  } catch (error) {
    logger.error(`Error creating session: ${error.message}`);
    const errorResponse = new ErrorResponse(
      req.body.id || uuidv4(),
      'internal_error',
      'An unexpected error occurred'
    );
    return res.status(500).json(errorResponse);
  }
}

/**
 * Text completion endpoint - Streamable HTTP
 */
function getTextCompletions(req, res) {
  try {
    const request = req.body;
    logger.info(`Text completion requested for session ${request.session_id}`);
    
    // Validate session
    const sessionInfo = sessionService.getSession(request.session_id);
    if (!sessionInfo) {
      logger.warn(`Invalid session ID: ${request.session_id}`);
      const errorResponse = new ErrorResponse(
        request.id,
        'invalid_session',
        'Invalid or expired session ID'
      );
      return res.status(400).json(errorResponse);
    }
    
    // Set up response for streaming
    res.setHeader('Content-Type', 'application/x-ndjson');
    
    // Generate demo responses
    const responses = generateDemoResponses(request);
    
    // Stream responses with small delays to simulate generation
    let index = 0;
    const interval = setInterval(() => {
      if (index < responses.length) {
        const json = JSON.stringify(responses[index]) + '\n';
        res.write(json);
        index++;
      } else {
        clearInterval(interval);
        res.end();
      }
    }, 100);
    
    // Handle client disconnect
    req.on('close', () => {
      clearInterval(interval);
      logger.info(`Client disconnected from text completion stream for session ${request.session_id}`);
    });
  } catch (error) {
    logger.error(`Error processing text completion: ${error.message}`);
    const errorResponse = new ErrorResponse(
      req.body.id || uuidv4(),
      'internal_error',
      'An unexpected error occurred'
    );
    return res.status(500).json(errorResponse);
  }
}

/**
 * Legacy HTTP+SSE streaming endpoint
 */
function streamTextCompletions(req, res) {
  try {
    const request = req.body;
    logger.info(`SSE text completion requested for session ${request.session_id}`);
    
    // Validate session
    const sessionInfo = sessionService.getSession(request.session_id);
    if (!sessionInfo) {
      logger.warn(`Invalid session ID: ${request.session_id}`);
      const errorResponse = new ErrorResponse(
        request.id,
        'invalid_session',
        'Invalid or expired session ID'
      );
      return res.status(400).json(errorResponse);
    }
    
    // Set up SSE headers
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    
    // Generate demo responses
    const responses = generateDemoResponses(request);
    
    // Stream responses with small delays to simulate generation
    let index = 0;
    const interval = setInterval(() => {
      if (index < responses.length) {
        const json = JSON.stringify(responses[index]);
        res.write(`data: ${json}\n\n`);
        index++;
      } else {
        res.write('data: [DONE]\n\n');
        clearInterval(interval);
        res.end();
      }
    }, 100);
    
    // Handle client disconnect
    req.on('close', () => {
      clearInterval(interval);
      logger.info(`Client disconnected from SSE stream for session ${request.session_id}`);
    });
  } catch (error) {
    logger.error(`Error processing SSE text completion: ${error.message}`);
    const errorResponse = new ErrorResponse(
      req.body.id || uuidv4(),
      'internal_error',
      'An unexpected error occurred'
    );
    return res.status(500).json(errorResponse);
  }
}

/**
 * Close session endpoint
 */
function closeSession(req, res) {
  try {
    const sessionId = req.params.sessionId;
    logger.info(`Session close requested: ${sessionId}`);
    
    // Try to close the session
    if (sessionService.closeSession(sessionId)) {
      const response = new SessionCloseResponse(uuidv4());
      
      logger.info(`Session closed successfully: ${sessionId}`);
      return res.json(response);
    } else {
      logger.warn(`Failed to close nonexistent session: ${sessionId}`);
      const errorResponse = new ErrorResponse(
        uuidv4(),
        'invalid_session',
        'Session not found'
      );
      return res.status(404).json(errorResponse);
    }
  } catch (error) {
    logger.error(`Error closing session: ${error.message}`);
    const errorResponse = new ErrorResponse(
      uuidv4(),
      'internal_error',
      'An unexpected error occurred'
    );
    return res.status(500).json(errorResponse);
  }
}

/**
 * Generate demo responses for text completion
 */
function generateDemoResponses(request) {
  // Generate simulated responses for demonstration purposes
  const lorem = "This is a demonstration of the Model Context Protocol server implemented in Node.js. " +
               "The server supports both modern Streamable HTTP and legacy HTTP+SSE transport options. " +
               "In a real implementation, this would connect to an AI model to generate responses based on the provided prompt.";
  
  const words = lorem.split(' ');
  const responses = [];
  
  // Initial chunks with content
  for (let i = 0; i < words.length; i++) {
    responses.push(new TextCompletionStreamResponse(
      request.id,
      new TextCompletionDelta(words[i] + ' ')
    ));
  }
  
  // Final chunk with finish reason and usage
  responses.push(new TextCompletionStreamResponse(
    request.id,
    new TextCompletionDelta('', 'stop'),
    new TextCompletionUsage(
      Math.ceil(request.inputs.prompt.length / 4), // Very rough approximation
      Math.ceil(lorem.length / 4),
      Math.ceil((request.inputs.prompt.length + lorem.length) / 4)
    )
  ));
  
  return responses;
}

module.exports = {
  createSession,
  getTextCompletions,
  streamTextCompletions,
  closeSession
};